db.go

```diff
@@ -485,110 +477,191 @@ func (db *DB) reload() (err error) {
+	}
+
+	// Swap new blocks first for subsequently created readers to be seen.
+	db.mtx.Lock()
+	oldBlocks := db.blocks
+	db.blocks = loadable
+	db.mtx.Unlock()
+
+	for _, b := range oldBlocks {
+		if _, ok := deletable[b.Meta().ULID]; ok {
+			deletable[b.Meta().ULID] = b
 		}
 	}
-	// Load new blocks into memory.
+
+	if err := db.deleteBlocks(deletable); err != nil {
+		return err
+	}
+
+	// Garbage collect data in the head if the most recent persisted block
+	// covers data of its current time range.
+	if len(loadable) == 0 {
+		return nil
+	}
+
+	maxt := loadable[len(loadable)-1].Meta().MaxTime
+
+	return errors.Wrap(db.head.Truncate(maxt), "head truncate failed")
+}
+
+func (db *DB) openBlocks() (blocks []*Block, corrupted map[ulid.ULID]error, err error) {
+	corrupted = make(map[ulid.ULID]error)
+
+	dirs, err := blockDirs(db.dir)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "find blocks")
+	}
+
 	for _, dir := range dirs {
-		meta, err := readMetaFile(dir)
+		ulid, err := ulid.Parse(filepath.Base(dir))
 		if err != nil {
-			return errors.Wrapf(err, "read meta information %s", dir)
-		}
-		// Don't load blocks that are scheduled for deletion.
-		if _, ok := deleteable[meta.ULID]; ok {
+			level.Error(db.logger).Log("msg", "not a block dir", "dir", dir)
 			continue
 		}
+
 		// See if we already have the block in memory or open it otherwise.
-		b, ok := db.getBlock(meta.ULID)
+		block, ok := db.getBlock(ulid)
 		if !ok {
-			b, err = OpenBlock(dir, db.chunkPool)
+			block, err = OpenBlock(db.logger, dir, db.chunkPool)
 			if err != nil {
-				return errors.Wrapf(err, "open block %s", dir)
+				corrupted[ulid] = err
 			}
+
 		}
-		blocks = append(blocks, b)
-		opened[meta.ULID] = struct{}{}
+		blocks = append(blocks, block)
 	}
+
+	return blocks, corrupted, nil
+}
+
+// deletableBlocks returns all blocks past retention policy
+// and blocks superseded by parents after a compaction.
+func (db *DB) deletableBlocks(blocks []*Block) map[ulid.ULID]*Block {
+	deletable := make(map[ulid.ULID]*Block)
+
+	// Delete old blocks that have been superseded by new ones by gathering their parents.
+	// Those parents all have newer replacements and
+	// can be safely deleted after loading the other blocks.
+	// This makes it resilient against the process crashing towards the end of a compaction.
+	// Creation of a new block and deletion of its parents cannot happen atomically.
+	// By creating blocks with their parents, we can pick up the deletion where it left off during a crash.
+	for _, block := range blocks {
+		for _, b := range block.Meta().Compaction.Parents {
+			deletable[b.ULID] = nil
+		}
+	}
+
+	// Sort the blocks by time - newest to oldest (largest to smallest timestamp).
+	// This ensures that the retentions will remove the oldest  blocks.
 	sort.Slice(blocks, func(i, j int) bool {
-		return blocks[i].Meta().MinTime < blocks[j].Meta().MinTime
+		return blocks[i].Meta().MaxTime > blocks[j].Meta().MaxTime
```

Comment:
Why was this changed? Curious if it changes anything....

Reply:
I really can't remember but reverting it fails the TimeRetention test